developer-guide.md

```diff
@@ -0,0 +1,546 @@
+
+### Example: Insert Multiple Documents
+
+In this example, we use the `bulk` helper function to insert multiple documents at once.
+
+Helpers also provide a `parallel_bulk` function that can be used to insert documents in parallel and a `streaming_bulk` function that can be used to insert documents in a streaming fashion.
+
+Read more about [helpers](https://elasticsearch-py.readthedocs.io/en/stable/helpers.html).
+
+```python
+from elasticsearch.helpers import bulk
+
+docs = [
+  { "text": [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5], "title": "Document 1"},
+  { "text": [0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4], "title": "Document 2"},
+  { "text": [0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3], "title": "Document 3"}
+]
+
+actions = []
+for doc in docs:
+  action = {
+      "_index": "demo",
+      "_source": doc
+  }
+  actions.append(action)
+
+bulk(es, actions, refresh=True)
+
+```
+
+## Querying Documents
+
+Once you have indexed documents, you can search for them using the `search` API.
+
+### Example: Query for Similar Documents
+
+In this example, we search for similar documents to the query vector.
+
+```python
+
+response = es.search(index="demo", body={
+    "knn": {
+      "field": "text",
+      "query_vector": [0.5, 0.5, 0.5, 0.5, 0.5, 0.5],
+      "k": 10, # return top 10 most similar documents
+      "num_candidates": 100
+    }
+})
+
+```
+
+### Example: Using model in Elasticsearch
+
+In this example, instead of passing the query vector, we pass the query text and Elasticsearch will use the embedding model deployed in Elasticsearch to transform the text into a vector.
+
+To deploy the model in Elasticsearch, follow the instructions in the [Deploying a Model](#deploying-a-model) section.
+
+```python
+
+response = es.search(index="demo", body={
+    "knn": {
+      "field": "text",
+      "k": 10,
+      "num_candidates": 50,
+      "query_vector_builder": {
+        "text_embedding": {
+          "model_id": "sentence-transformers__all-minilm-l6-v2", # model id deployed in Elasticsearch
+          "model_text": "example query"
+        }
+      }
+    }
+  }
+)
+
+```
+
+# Data Modeling
+
+Depending on the type of use-case you are building, you will need to model your data differently.
+
+## Long unstructured text
+
+If each document contains long text, you should consider breaking up the text into smaller chunks called passages. Long text are hard to represent as a single vector, especially if the text is about multiple topics. We recommend breaking up the text into smaller chunks, and indexing each passage as a separate document.
+
+Breaking up the text into passages allows you to retrieve more relevant results, as you can find the most relevant passages in one or more documents that match the query.
+
+Another advantage is for RAG use-cases, being able to pass in smaller relevant passages to the RAG model, instead of the entire document, will improve the quality of the answer and avoid hitting the token limits of the RAG model.
```

Comment:
```suggestion
Another advantage is for Retrieval Augmented Generation (RAG) use-cases. Passing smaller, relevant passages to the model, instead of the entire document, will improve the quality of the answer and avoid hitting token limits.
```